File: datasets/Part 2 - Regression/decission_tree_regression.R

# Arbol de Decision para Regresion

setwd("~/repos/machinelearning-az/datasets/Part 2 - Regression")

# Importar el dataset
dataset = read.csv('data/Position_Salaries.csv')
dataset = dataset[, 2:3]

# En este caso no se va a dividir en conjunto de entrenamiento y test
# debido a la escasa cantidad de datos

# Dividir los datos en conjunto de entrenamiento y conjunto de test
# library(caTools) # Instalar con install.packages('caTools')
# set.seed(123) # Para que salgan los mismos resultados que en el curso
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)


# Escalado de valores
#training_set[, 2:3] = scale(training_set[, 2:3])
#testing_set[, 2:3] = scale(testing_set[, 2:3])


# Ajustar Modelo de Regresion con el Conjunto de Datos
library(rpart) # Instalar con install.packages('rpart')
regression = rpart(formula = Salary ~ .,
                   data = dataset,
                   control = rpart.control(minsplit = 1))

# Prediccion de nuevos resultados con Arbol de Regresion
y_pred = predict(regression, newdata = data.frame(Level = 6.5))

# Visualizacion del Modelo de Arbol de Regresion
library(ggplot2) # Instalar con install.packages('ggplot2')
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = x_grid, y = predict(regression,
                                        newdata = data.frame(Level = x_grid))),
            color = 'blue') +
  ggtitle('Prediccion con Arbol de Decision (Modelo de Regresion)') +
  xlab('Nivel del empleado') +
  ylab('Sueldo (en $)')

File: datasets/Part 2 - Regression/decission_tree_regression.py
# -*- coding: utf-8 -*-
"""
Created on Sun May  8 12:54:08 2022

@author: mudar
"""

# Regresion con Arboles de Decision

# Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('data/Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values # para que no sea un vector y sea una matriz ponemos 1:2
y = dataset.iloc[:, 2].values


# Dividir el dataset en conjunto de entrenamiento y conjunto de testing
# 80% conjunto entrenamiento y 20% conjunto de testing
"""
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
"""


# Escalado de variables
# Para que no haya tanta diferencia en los rangos de
# valores en los datos de las distintas columnas
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""


# Ajustar la Regresion con todo el dataset
from sklearn.tree import DecisionTreeRegressor
regression = DecisionTreeRegressor(random_state = 0)
regression.fit(X, y)


# Prediccion de nuestro modelo de regresion con Arboles de Decision
y_pred = regression.predict([[6.5]])


# Visualizacion de los resultados del Modelo
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'red')
plt.plot(X, regression.predict(X), color = 'blue')
plt.title('Modelo de Regresion con Arboles de Decision')
plt.xlabel('Posicion del empleado')
plt.ylabel('Sueldo (en $)')
plt.show()
